### ISSUE TITLE
Disable "expression value is unused" for specific classes (motivating example: Airflow operator DSL)

### ISSUE THREAD

**Issue by Throne3d** [microsoft/pyright#6612]
**Is your feature request related to a problem? Please describe.**

When enabling stricter type checking, "expression value is unused" appears on the result of a right shift operation - this makes sense! However, in the case of [Apache Airflow](https://github.com/apache/airflow)'s DSL, it overrides the lshift + rshift operations on the Operator and Sensor classes to define dependencies between items in a DAG, and I see a lot of code that triggers this "expression value is unused" warning as a result. So far I've been suppressing this warning in several projects to reduce warning fatigue here, but it either means a lot of `type: ignore` / `pyright: ignore[reportUnusedExpression]` over these (several) lines, or disabling the diagnostic for whole files/projects, which means we lose the benefit of the diagnostic catching more subtle bugs.

**Describe the solution you’d like**

All operators + sensors in Airflow eventually inherit from [a DependencyMixin class with the overrides](https://github.com/apache/airflow/blob/2.7.3/airflow/models/taskmixin.py#L96). It would be great to be able to disable this warning when the rshift/lshift (or rrshift/rlshift) operation is applied to a class that inherits from DependencyMixin. To generalize this feature to other situations, it would be good to allow disabling this configuration for specific classes (ideally tracking inheritance here), possibly for specific operations instead of all of them, as a configuration defined in tool.pyright. This idea came up in #5264 (disabling the check on specific classes) but I think it got lost.

**Examples**

The code can be rewritten to use `.set_downstream()` or `.set_upstream()` instead, but I rarely see code use that, and it can make chained dependencies a lot longer and in some cases harder to understand. Example code might look like:

```py
import datetime
from airflow import DAG
from airflow.operators.empty import EmptyOperator
from airflow.sensors.time_delta import TimeDeltaSensorAsync

with DAG(dag_id="sample-dag") as dag:
    wait = TimeDeltaSensorAsync(task_id="wait", delta=datetime.timedelta(seconds=30))
    finish = EmptyOperator(task_id="finish")

    wait >> finish  # Expression value is unused
```

For an example of chained and branched dependencies, where this rewrite can make these dependencies less clear:

```py
    task1 >> task2 >> task3 >> task4
    # becomes
    task1.set_downstream(task2)
    task2.set_downstream(task3)
    task3.set_downstream(task4)

    task1 >> [task2a, task2b] >> task3 >> task4
    # becomes
    task1.set_downstream([task2a, task2b])  # equivalent to task1.set_downstream(task2a) and task1.set_downstream(task2b)
    task3.set_upstream([task2a, task2b])
    task3.set_downstream(task4)
```

Thanks!

**Comment by hmc-cs-mdrissi** [comment]
I'd suggest [here](https://discuss.python.org/c/typing/32). I'd expect if there was a mechanism to mark a binary operator as impure it would need some typing community discussion first.

I'm supportive of this idea as I can think of couple other libraries (dataflow) that have similar behavior where they rely on some operators to define DAG for execution. But I think pyright isn't best place to start discussion and having multiple library examples would be good place to start if you want to propose this as a new typing feature.

One simple way would be to add a decorator like `@impure` that indicates a function has side effects and should not trigger unused checks. Something like,

```python
class Foo:
  @impure
  def __rshift__(self, other: Foo) -> Foo:
    ...
```

**Comment by erictraut** [comment]
The `reportUnusedExpression` rule in pyright is a strict-mode check and is somewhat opinionated in how it works. It is disabled by default, and you don't need to enable it if it doesn't serve your purpose. You can also disable it on a per-file basis (using a `# pyright: reportUnusedExpression = false` comment at the top of the file) if that helps.

I personally don't think it's a good idea for libraries to overload binary operators in a way that is inconsistent with the conventional meaning of that operator, but if you happen to use a library that does so, you may need to leave this particular type checking feature disabled.

Pyright has no knowledge of any particular third-party libraries, and it's unlikely we'd ever embed such special casing for a particular class in a particular library.

If there is sufficient need for a standardized mechanism for exempting a particular binary (or unary) operator override from this check, then the Python typing forum would be the place to start that discussion, as mentioned by @hmc-cs-mdrissi.
